packages/interactivity/src/proxies/store.ts

```diff
@@ -0,0 +1,56 @@
+/**
+ * Internal dependencies
+ */
+import { createProxy, getProxyNs, shouldProxy } from './registry';
+import { setNamespace, resetNamespace } from '../hooks';
+import { withScope } from '../utils';
+
+const isObject = ( item: unknown ): item is Record< string, unknown > =>
+	Boolean( item && typeof item === 'object' && item.constructor === Object );
+
+const storeRoots = new WeakSet();
+
+const storeHandlers: ProxyHandler< object > = {
+	get: ( target: any, key: string | symbol, receiver: any ) => {
+		const result = Reflect.get( target, key );
+		const ns = getProxyNs( receiver );
+
+		// Check if the proxy is the store root and no key with that name exist. In
+		// that case, return an empty object for the requested key.
+		if ( typeof result === 'undefined' && storeRoots.has( receiver ) ) {
+			const obj = {};
+			Reflect.set( target, key, obj );
+			return proxifyStore( ns, obj, true );
+		}
+
+		// Check if the property is a function. If it is, add the store
+		// namespace to the stack and wrap the function with the current scope.
+		// The `withScope` util handles both synchronous functions and generator
+		// functions.
+		if ( typeof result === 'function' ) {
+			setNamespace( ns );
+			const scoped = withScope( result );
+			resetNamespace();
+			return scoped;
+		}
+
+		// Check if the property is an object. If it is, proxyify it.
+		if ( isObject( result ) && shouldProxy( result ) ) {
+			return proxifyStore( ns, result, true );
+		}
+
+		return result;
+	},
+};
+
+export const proxifyStore = < T extends object >(
+	namespace: string,
+	obj: T,
+	isNotRoot = false
+): T => {
+	const proxy = createProxy( namespace, obj, storeHandlers );
+	if ( proxy && ! isNotRoot ) {
```

Comment:
Could we call the parameter `isRoot` and then change the `&& ! isNotRoot` to `&& isRoot` ?

Reply:
Actually, that was my first approach. I just changed it because I felt the default case would be creating a root object. I can change the value to be `true` by default, though. 🙂

Reply:
Let's avoid the variable names with negations, as it's hard to reason about `! isNotRoot` in contrast to `isRoot`.

Reply:
Fixed in https://github.com/WordPress/gutenberg/pull/62734/commits/5c17dd76651fb88620a4c48b4eb4ad45a6f11f5f.